File: 438.find-all-anagrams-in-a-string.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        Map<Character, Integer> map = new HashMap<Character, Integer>();
        for (char c : p.toCharArray()) {
            map.put(c, map.getOrDefault(c, 0)+1);
        }
        List<Integer> res = new ArrayList<Integer>();
        Map<Character, Integer> countMap = new HashMap<Character, Integer>();
        int count = 0;
        for (int i = 0, j = 0; j < s.length(); j++) {
            char c = s.charAt(j);
            countMap.put(c, countMap.getOrDefault(c, 0)+1);
            // if (countMap.get(c) > 9999) {
            // System.out.println(countMap.get(c));
            // System.out.println(map.getOrDefault(c, 0));
            // }
            if (countMap.get(c).intValue() == map.getOrDefault(c, 0).intValue()) {
                count++;
            }
            if (j-i+1 == p.length()) {
                if (count == map.size()) {
                    res.add(i);
                }
                char cs = s.charAt(i);
                if (countMap.get(cs).intValue() == map.getOrDefault(cs, 0).intValue()) {
                    count--;
                }
                countMap.put(cs, countMap.get(cs)-1);
                i++;
            }
        }
        return res;
    }
}

File: 438.find-all-anagrams-in-a-string.go
func findAnagrams(s string, p string) []int {
	m := make(map[byte]int)
	for _, c := range p {
		m[byte(c)]++
	}
	res := []int{}
	countMap := make(map[byte]int)
	count := 0
	for i, j := 0, 0; j < len(s); j++ {
		countMap[s[j]]++
		if countMap[s[j]] == m[s[j]] {
			count++
		}
		if j-i+1 == len(p) {
			if count == len(m) {
				res = append(res, i)
			}
			if countMap[s[i]] == m[s[i]] {
				count--
			}
			countMap[s[i]]--
			i++
		}
	}
	return res
}
